encoding/thrift/inbound_nowire.go

```diff
@@ -0,0 +1,160 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package thrift
+
+import (
+	"context"
+	"io"
+
+	"go.uber.org/thriftrw/protocol/binary"
+	"go.uber.org/thriftrw/protocol/stream"
+	"go.uber.org/thriftrw/wire"
+	encodingapi "go.uber.org/yarpc/api/encoding"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/pkg/errors"
+)
+
+var _emptyResponse = NoWireResponse{}
+
+type thriftNoWireHandler struct {
+	NoWireHandler NoWireHandler
+	Protocol      stream.Protocol
+	Enveloping    bool
+}
+
+type NoWireCall struct {
+	Reader        io.Reader
+	RequestReader stream.RequestReader
+	EnvelopeType  wire.EnvelopeType
+
+	StreamReader stream.Reader
+}
+
+type NoWireHandler interface {
+	Handle(context.Context, *NoWireCall) (NoWireResponse, error)
+}
+
+func (t thriftNoWireHandler) Handle(ctx context.Context, treq *transport.Request, rw transport.ResponseWriter) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	defer closeReader(treq.Body)
+
+	res, err := doRequest(ctx, call, treq, rw, t.NoWireHandler, wire.Call, t.Protocol, t.Enveloping)
+	if err != nil {
+		return err
+	}
+
+	if resType := res.Body.EnvelopeType(); resType != wire.Reply {
+		return errors.ResponseBodyEncodeError(
+			treq, errUnexpectedEnvelopeType(resType))
+	}
+
+	if res.IsApplicationError {
+		rw.SetApplicationError()
+
+		if applicationErrorMetaSetter, ok := rw.(transport.ApplicationErrorMetaSetter); ok {
+			applicationErrorMetaSetter.SetApplicationErrorMeta(&transport.ApplicationErrorMeta{
+				Details: res.ApplicationErrorDetails,
+				Name:    res.ApplicationErrorName,
+				Code:    res.ApplicationErrorCode,
+			})
+		}
+	}
+
+	if err := call.WriteToResponse(rw); err != nil {
+		// not reachable
+		return err
+	}
+
+	if err := res.ResponseWriter.WriteResponse(wire.Reply, rw, res.Body); err != nil {
+		return errors.ResponseBodyEncodeError(treq, err)
+	}
+
+	return nil
+}
+
+func (t thriftNoWireHandler) HandleOneway(ctx context.Context, treq *transport.Request) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	defer closeReader(treq.Body)
+
+	_, err := doRequest(ctx, call, treq, nil, t.NoWireHandler, wire.OneWay, t.Protocol, t.Enveloping)
+	return err
+}
+
+func doRequest(
```

Comment:
Let's call this `decodeNoWireRequest`. Please add same level of comment we have in `decodeRequest` for this function.